Add a blog post about secure MCP SSE server (#2284)
quarkusio/quarkusio.github.io: _posts/2025-04-24-secure-mcp-sse-server.adoc

---
layout: post
title: 'Getting ready for secure MCP with Quarkus MCP Server'
date: 2025-04-24
tags: ai mcp security
synopsis: 'Explain how MCP clients can access secure Quarkus MCP SSE servers with access tokens'
author: sberyozkin
---
:imagesdir: /assets/images/posts/secure_mcp_sse_server

== Introduction

> https://modelcontextprotocol.io/specification/2025-03-26[The latest version of the Model Context Protocol (MCP) specification] introduces an https://modelcontextprotocol.io/specification/2025-03-26/basic/authorization[authorization] flow.

Reviewer: ```suggestion
```

I would just leave this out or put link somewhere below.

> While it will take a bit of time for the MCP authorization part of the MCP specification be finalized and widely supported, one thing you can be certain about is that MCP authorization compliant clients will be able to login users with the OAuth2 authorization code flow and use bearer tokens to access MCP servers on their behalf.

Author: @cescoffier May be I'm missing something,

>you can already add authentication to client and server following the previous MCP version (Add version date).

Where exactly in the previous MCP version, at https://modelcontextprotocol.io/specification/2024-11-05, one can do the authentication ?

> While it will take a bit of time for the MCP authorization part of the MCP specification be finalized and widely supported, one thing you can be certain about is that MCP authorization compliant clients will be able to login users with the OAuth2 authorization code flow and use bearer tokens to access MCP servers on their behalf.

Author: Here I'm trying to prepare users for what is coming with the `latest` spec, i.e, no matter how the MCP client acquires the token (and it must be authorization code flow according to the latest spec, with client_credentials mentioned but omitted from the Microsoft PR), MCP server will receive a bearer token

> While it will take a bit of time for the MCP authorization part of the MCP specification be finalized and widely supported, one thing you can be certain about is that MCP authorization compliant clients will be able to login users with the OAuth2 authorization code flow and use bearer tokens to access MCP servers on their behalf.

Reviewer: the old mcp client protocol is http based which support setting these headers - hence it works with existing one.

> While it will take a bit of time for the MCP authorization part of the MCP specification be finalized and widely supported, one thing you can be certain about is that MCP authorization compliant clients will be able to login users with the OAuth2 authorization code flow and use bearer tokens to access MCP servers on their behalf.

Reviewer 2: @sberyozkin can you apply my suggestion?

> While it will take a bit of time for the MCP authorization part of the MCP specification be finalized and widely supported, one thing you can be certain about is that MCP authorization compliant clients will be able to login users with the OAuth2 authorization code flow and use bearer tokens to access MCP servers on their behalf.

Author: Sounds good, @cescoffier, @maxandersen, indeed, it was already HTTP based before.

How does it look now ?

> In the meantime, you can use any MCP client that can accept a bearer access token and send it to the MCP server to start experimenting with MCP authorization.

Author: @cescoffier 

>This flow is quite similar to what is defined in the latest MCP specification (add link)

Unfortunately it is not. The latest specification describes an authorization code flow, just having some tool like MCP inspector be able to pass the token is not related to the specification as such

> In the meantime, you can use any MCP client that can accept a bearer access token and send it to the MCP server to start experimenting with MCP authorization.

Author: @cescoffier But that said, I can try to clarify that passing the bearer token is part of the latest MCP authprization proposal, thanks

> In the meantime, you can use any MCP client that can accept a bearer access token and send it to the MCP server to start experimenting with MCP authorization.

Author: @cescoffier I'll add a link to https://modelcontextprotocol.io/specification/2025-03-26/basic/authorization#2-6-access-token-usage

In this post, we will create a https://github.com/quarkiverse/quarkus-mcp-server[Quarkus MCP SSE Server] that requires authentication for the main and tools SSE endpoints, and test it with `curl`, https://modelcontextprotocol.io/docs/tools/inspector[MCP inspector], and `Quarkus MCP SSE Server Dev UI`.

Please review the https://modelcontextprotocol.io/specification/2025-03-26[latest MCP specification] for a complete explanation of the MCP architecture, core concepts, transports.

== Create MCP SSE server

> First, let's create a secure Quarkus MCP SSE server that requires authentication during the initial SSE handshake and the tool access.

Reviewer 2: ```suggestion
First, let's create a secure Quarkus MCP SSE server that requires authentication during the initial Server-Sent Events (SSE) handshake and the tool access.
```

</dependency>
----
<1> `quarkus-mcp-server-sse` is required to support the MCP SSE transport.
<2> `quarkus-oidc` is required to secure access to the MCP SSE endpoints.

[[tool]]
=== Tool

Let's create a tool that can be invoked only if the current MCP request is authenticated:

[source,java]
----
package org.acme;

import io.quarkiverse.mcp.server.TextContent;
import io.quarkiverse.mcp.server.Tool;
import io.quarkus.security.Authenticated;
import io.quarkus.security.identity.SecurityIdentity;
import jakarta.inject.Inject;

public class ServerFeatures {

    @Inject
    SecurityIdentity identity;

    @Tool(name = "user-name-provider", description = "Provides a name of the current user") <1>
    @Authenticated <2>
    TextContent provideUserName() {
        return new TextContent(identity.getPrincipal().getName()); <3>
    }
}
----
<1> Provide a tool that can return a name of the current user. Note the `user-name-provider` tool name, you will use it later for a tool call.
<2> Require authenticated tool access. See also how the main MCP SSE endpoint is secured in the <<configuration>> section below.
<3> Use the injected `SecurityIdentity` to return the current user's name.

> 

Author: @cescoffier Hi, I've updated the second point which will now read: `Require authenticated tool access - yes, the only difference with an unauthenticated MCP server tool is `@Authenticated`, that's it!.` - thanks; and indeed, the sentence that follows there refers to also authenticating the main sse endpoint.
Indeed, we'd need to make that one public to support fixing authenticated and unauthenticated tool calls

> 

Author: I.e, just `@Athenticated` alone protects the tool only

The OIDC configuration is provided in devmode by https://quarkus.io/guides/security-openid-connect-dev-services[Dev Services for Keycloak]. It creates a default realm, client and adds two users, `alice` and `bob`, for you to get started immediately. You can also register a custom Keycloak realm to work with the the existing realm, client and user registrations.

No problems if you do not work with Keycloak, see the <<mcp-server-devui>> section for more details.

We are ready to test our secure MCP server.

== Start MCP SSE server

Start the server in the dev mode:

[source,shell]
----
mvn quarkus:dev
----

== Access MCP SSE server

[[mcp-server-devui]]
=== Use Quarkus MCP Server Dev UI

Go to http://localhost:8080/q/dev[Dev UI], find both MCP Server and OpenId Connect cards:

image::mcp_server_oidc_devui.png[MCP Server and OIDC in DevUI,align="center"]

> Select an OpenId Connect card and https://quarkus.io/guides/security-openid-connect-dev-services#develop-service-applications[login to Keycloak] using an `alice` name and an `alice` password.

Reviewer 2: ```suggestion
Select the OpenId Connect card and https://quarkus.io/guides/security-openid-connect-dev-services#develop-service-applications[login to Keycloak] using an `alice` name and an `alice` password.
```

I'm not sure _select_ is the right word

Author: Sure, I'll just use `choose`

[NOTE]
====
You can login to other providers such as `Auth0` or https://quarkus.io/guides/security-openid-connect-providers#github[GitHub] from OIDC DevUI as well. The only requirement is to update your application registration to allow callbacks to DevUI. For example, see how you can https://quarkus.io/guides/security-oidc-auth0-tutorial#looking-at-auth0-tokens-in-the-oidc-dev-ui[login to Auth0 from Dev UI].

> 

Reviewer 2: ```suggestion
===
```

otherwise, you are still in the admonition.

After logging in with `Keycloak` as `alice`, copy the acquired access token using a provided copy button:

image::login_and_copy_access_token.png[Login and copy access token,align="center"]

Next, go to the MCP Server card, select the `Tools` option and choose to `Call` the `user-name-provider` tool:

image::mcp_server_choose_tool.png[Choose MCP Server tool,align="center"]

Paste the copied access token into the Tool's `Bearer token` field, and request a new MCP SSE session:

image::mcp_server_bearer_token.png[MCP Server Bearer token,align="center"]

Use the `Call` action to make a tool call and get a response which contains the `alice` user name:

image::mcp_server_tool_response.png[MCP Server tool response,align="center"]

> 

Reviewer 2: It's great, but I'm wondering if a short screen recording would not be better (without voice over).

Author: Hi @cescoffier I propose to do it in follow up presentations, or may be in Insights, I'll have to refresh my knowledge again how to do recordings on Fedora. I did it before, but the software is not on the bew laptop.

[[mcp-inspector]]

> === Use MCP Inspector

Reviewer 2: == Step 3 - Invoking an MCP tool requiring authentication with the MCP Inspector

> 

Reviewer 2: ```suggestion
The [MCP Inspector](https://github.com/modelcontextprotocol/inspector) is an interactive developer utility for testing and debugging MCP servers. It is handy to understand what's going on. Let's use this inspector to invoke our MCP server with authentication.
```

Launch https://modelcontextprotocol.io/docs/tools/inspector[MCP inspector]:

[source,shell]
----
npx @modelcontextprotocol/inspector
----

> Next, get and copy the access token as explained in the <<mcp-server-devui>> section above. You can use the same access token that you may have already used when working through that section.

Reviewer 2: Here I would do things slightly differently. I would run the application in "prod" mode - no dev UI, _real_ keycloak. Do you think it would be possible?

Author: @cescoffier I may have missed it, but yeah, now we do it with GitHub

Use this access token to have MCP inspector connect to the Quarkus MCP SSE server:

image::mcp_inspector_connect.png[MCP Inspector Connect,align="center"]

Next, make a `user-name-provider` tool call:

image::mcp_inspector_tool_call.png[MCP Inspector Tool Call,align="center"]

> === Use curl

Reviewer 2: ```suggestion
== == Step 4 - Invoking an MCP tool requiring authentication with curl
```

Finally, let's use `curl` and also learn a little bit how the MCP protocol and MCP SSE transport work. 

> 

Reviewer 2: What about adding something like this (and a picture):

When invoking an MCP tool using the SSE transport, there are at least two requests:

1. A first request connects to the server using SSE - this _back channel_ receives the invocation results and notifications
2. A second request (`POST`) invokes the tool. The invocation result is transmitted on the SSE _back channel_.

Author: @cescoffier Sorry, missed it too, good idea, if you still think it is worth adding it, please create a diagram with your nice tools, may be on Monday ? and We  can add it to the `curl` section, I can try to sketch it as well, but you have a very clear picture how it should be presented, so yeah, please do it and we can add it tomorrow

> First, access the main SSE endpoint without the access token:

Reviewer 2: Same if possible I would use prod mode. I think it's important to show that this feature is not only a _dev_ thing, it's a real deal.

[source,shell]
----
curl -v localhost:8080/mcp/sse
----

You will get HTTP 401 error.

Next, get and copy the access token as explained in the <<mcp-server-devui>> section above. You can use the same access token that you may have already used when working through that section.

[source,shell]
----
curl -v -H "Authorization: Bearer ey..." localhost:8080/mcp/sse
----

and get an SSE response such as:

[source,shell]
----
< content-type: text/event-stream
< 
event: endpoint
data: /messages/ZTZjZDE5MzItZDE1ZC00NzBjLTk0ZmYtYThiYTgwNzI1MGJ
----

The current `curl` SSE session must stay open, for it to receive updates related to the dynamically allocated `data` endpoint.

Open another window and use the same access token to post a tool call request to the `data` endpoint:

[source,shell]
----

> curl -v -H "Authorization: Bearer ey..." -H "Content-Type: application/json" --data @call.json http://localhost:8080/mcp/messages/ZTZjZDE5MzItZDE1ZC00NzBjLTk0ZmYtYThiYTgwNzI1MGJ

Reviewer 2: I know it will automatically switch to POST as you have a body, but would it be possible to specify the verb? Not everyone is a curl expert :-)

----

where the `call.json` file looks like this:

[source,json]
----
{
  "jsonrpc": "2.0",
  "id": 2,
  "method": "tools/call",
  "params": {
    "name": "user-name-provider",
    "arguments": {
    }
  }
}
----

> Now look in the first curl window and observe a response which contains the `alice` user name:

Reviewer 2: ```suggestion
Now look the SSE connection and observe a response which contains the `alice` user name:
```

[source,shell]
----
event: message
data: {"jsonrpc":"2.0","id":2,"result":{"isError":false,"content":[{"text":"alice","type":"text"}]}}
----

== Conclusion

> In this blog post, we have explained how you can easily create a secure Quarkus MCP SSE server, obtain a bearer access token and test it in 

Reviewer 2: ```suggestion
In this blog post, we have explained how you can easily create an authenticated Quarkus MCP SSE server, obtain a bearer access token and test it in 
```

Reviewer 2: in what? Something missing :-)

Author: @cescoffier Hi, lol, I probably jumped to some other section, sure, will fix
